xonsh/environ.py

```diff
@@ -7,62 +7,188 @@
+        # to the self._run_every set.
+        self['time24'] = functools.partial(time.strftime, '%H:%M:%S')
+        self._run_every.add('time24')
+
+    def login_time(self):
+        who_out = subprocess.check_output(['who', '-m'])
+        return ' '.join(who_out.decode().split()[2:4])
+    '''
+
+    def __init__(self, *args, **kwargs):
+
+        super(PromptFormatter, self).__init__(*args, **kwargs)
+        self._run_every = set()
+
+    def __getitem__(self, key):
+        '''
+        If the item is a function, return the output of the function.
+        This is so we can have static prompt values that only need to be
+        determined once and dynamic values that have to be recomputed each
+        time the prompt is displayed.
+        '''
+        value = super(PromptFormatter, self).__getitem__(key)
+
+        if callable(value):
+            if key in self._run_every:
+                # Variables that need to be computed every time
+                return value()
+            # Variables computed a single time
+            value = value()
+            self[key] = value
+
+        # If not callable then it was a static value
+        return value
 
-    prompt_scripts = [
-        '/usr/lib/git-core/git-sh-prompt',
-        '/usr/local/etc/bash_completion.d/git-prompt.sh'
-    ]
+    def get(self, key, default=None):
+        '''Return the value for the key.
 
-    for script in prompt_scripts:
-        # note that this is about 10x faster than bash -i "__git_ps1"
-        _input = ('source {}; __git_ps1 "${{1:-%s}}"'.format(script))
+        If the key is not present, return the value of the default param.
+        If the value is a function, return the output of the function.
+
+        This is implemented for completeness but isn't currently used
+        internally to xonsh.
+        '''
         try:
-            branch = subprocess.check_output(['bash',], cwd=cwd, input=_input,
-                                        stderr=subprocess.PIPE,
-                                        universal_newlines=True) or None
-        except subprocess.CalledProcessError:
-            continue
+            value = self[key]
+        except KeyError:
+            return default
+        return value
+
+
+class DefaultPromptFormatter(PromptFormatter):
+    def __init__(self, *args, **kwargs):
+        super(DefaultPromptFormatter, self).__init__(*args, **kwargs)
+        self._env = builtins.__xonsh_env__
+
+        self.update(TERM_COLORS)
+        formatters = dict(
+                base_cwd=partial(self.cwd, False),
+                cwd=self.cwd,
+                curr_branch=self.curr_branch,
+                hostname=socket.gethostname,
+                short_host=lambda: socket.gethostname().split('.', 1)[0],
+                time=datetime.now,
+                user=self.user,
+        )
+        self.update(formatters)
+        self._run_every.update(frozenset(('base_cwd', 'cwd', 'curr_branch', 'time')))
```

Comment:
Yeah, I was wondering about that.  When would users change during the lifetime of a shell?  I tried to think about sudo, su, and ssh.  It seemed like su and ssh both create a new shell.  sudo will run a program as another user but by the time you get your shell prompt back you'll be the same user you started as.  It also seemed to make a sort of conceptual sense in that if you became a new user and were able to keep interacting with your original shell the shell would need to have all sorts of authorization handling to prevent the second account from doing things as the first account.

If you can think of something that I've overlooked, I'm happy to add user in there, though.